cpp/include/graph.hpp

```diff
@@ -240,11 +237,188 @@ class GraphCSC: public GraphCompressedSparseBase<VT,ET,WT> {
+   * @param  number_of_edges       The number of edges in the graph
+   */
+  GraphCOO(VT number_of_vertices,
+           ET number_of_edges,
+           bool has_data = false):
+    number_of_vertices_(number_of_vertices),
+    number_of_edges_(number_of_edges_),
+    src_indices_(sizeof(VT)*number_of_edges),
+    dst_indices_(sizeof(VT)*number_of_edges)
+    edge_data_(has_data? sizeof(WT)*number_of_edges : 0)
+  {}
+
+  VT* src_indices(void) { return static_cast<VT*>(src_indices_.data()); }
+  VT* dst_indices(void) { return static_cast<VT*>(dst_indices_.data()); }
+  WT* edge_data(void) { return static_cast<WT*>(edge_data_.data()); }
+
+  struct contents {
+    std::unique_ptr<rmm::device_buffer> src_indices;
+    std::unique_ptr<rmm::device_buffer> dst_indices;
+    std::unique_ptr<rmm::device_buffer> edge_data;
+  };
+
+  contents release() noexcept;
+
+};
+
+/**
+ * @brief       Base class for constructted graphs stored in CSR (Compressed Sparse Row) format or CSC (Compressed Sparse Column) format
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class GraphCompressedSparseBase {
+  VT number_of_vertices_;
+  ET number_of_edges_;
+  rmm::device_buffer offsets_{};   ///< CSR offsets
+  rmm::device_buffer indices_{};   ///< CSR indices
+  rmm::device_buffer edge_data_{}; ///< CSR data
+
+public:
+
+  /**
+   * @brief      Take ownership of the provided graph arrays in CSR/CSC format
+   *
+   * @param  offsets               This array of size V+1 (V is number of vertices) contains the offset of adjacency lists of every vertex.
+   *                               Offsets must be in the range [0, E] (number of edges).
+   * @param  indices               This array of size E contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @param  edge_data             This array of size E (number of edges) contains the weight for each edge.  This
+   *                               array can be null in which case the graph is considered unweighted.
+   * @param  number_of_vertices    The number of vertices in the graph
+   * @param  number_of_edges       The number of edges in the graph
+   */
+  GraphCompressedSparseBase(VT number_of_vertices,
+                            ET number_of_edges,
+                            bool has_data):
+    number_of_vertices_(number_of_vertices),
+    number_of_edges_(number_of_edges_),
+    offsets_(sizeof(ET)*(number_of_vertices + 1)),
+    indices_(sizeof(VT)*number_of_edges),
+    edge_data_(has_data? sizeof(WT)*number_of_edges : 0)
   {}
+
+  ET* offsets(void) { return static_cast<ET*>(offsets_.data()); }
+  VT* indices(void) { return static_cast<VT*>(indices_.data()); }
+  WT* edge_data(void) { return static_cast<WT*>(edge_data_.data()); }
+
+  struct contents {
```

Comment:
I would assume the algorithms will be returning these contents conatining rmm device_buffers when they have to return a Graph. So in my current understanding an algorithm workflow could possibly look like this:

        Algorithm_xyz(py_Graph, py_subgraph):
            py_Graph.view_adj_list()     // view_adj_list() defined as a graph functionality [1]
            // Populating input Cpp Graph from py_graph adjlist data. Same as before... [2]
            num_verts = py_graph.number_of_vertices()
            num_edges = len(py_indices)
            cdef uintptr_t c_offsets = py_offsets.__cuda_array_interface__['data'][0]
            cdef uintptr_t c_indices = py_indices.__cuda_array_interface__['data'][0]
            cdef uintptr_t c_weights = <uintptr_t>NULL
            graph_float = GraphCSRView[int,int,float](<int*>c_offsets, <int*>c_indices, <float*>c_weights, num_verts, num_edges)
            // Algorithm returns a subgraph
            std::unique_ptr<GraphCOO> c_algorithm_xyz(graph_float)
            // Put subgraph data into py_subgraph [2]
            get_**contents**_from_GraphCOO     // similar to what cudf employs for their column 
            subgraph_offsets = DeviceBuffer.c_from_unique_ptr(contents.offsets)
            ....
            py_subgraph.adjlist.offsets = cudf.Series(Buffer(subgraph_offsets), dtype=np.int32)
            ....
            return py_subgraph

        view_adj_list(self):  // Graph function called in [1]
            Populate GraphCOOView from python Graph  // Similar to [2]
            call cpp function for adding adjlist that returns a GraphCSR
            Populate python Graph from GraphCSR // Same as [3]

Reply:
Yes, the classes have a function `contents release() noexcept;` which need to be called by the python layer. As you wrote in the example, this will get you access to the underlying buffers.